synapse/storage/prepare_database.py

```diff
@@ -150,6 +146,24 @@ def prepare_database(
                     % (SCHEMA_VERSION, version_info.current_version)
                 )
 
+            # Otherwise if we are on the current schema version, we need to check if there
+            # are any unapplied delta files to process. This should be run on all processes,
+            # master or worker. The master will apply the deltas, workers will check if any
+            # outstanding deltas exist and bail out if they do.
+            if version_info.current_version == SCHEMA_VERSION:
```

Comment:
we need to call `_upgrade_existing_database` to do the upgrade on the master.

I can't help but wonder if this wouldn't be simpler if we moved the logic into `_upgrade_existing_database`, since that's already got conditional code for workers.

Reply:
Good point on both. I've moved the logic (simplifying it quite a bit) into `_upgrade_existing_database` in 
5aa7589, and made sure the master process actually runs its migrations :)